### PR TITLE
Refactor due-diligence menu to new refactoring style

### PR DESCRIPTION
**Pull request by DidierRLopes**

### CODE REVIEW

**Comment by DidierRLopes** [comment]
> So looking at the yaml:
> 
> ```yaml
> interactions:
> - request:
>     body: null
>     headers:
>       Connection:
>       - close
>       Host:
>       - financialmodelingprep.com
>       User-Agent:
>       - Python-urllib/3.8
>     method: GET
>     uri: https://financialmodelingprep.com/api/v3/historical-rating/PLTR?apikey=REPLACE_ME
>   response:
>     body:
>       string: '{"Error Message" : "Invalid API KEY. Please retry or visit our documentation
>         to create one FREE https://financialmodelingprep.com/developer/docs"}'
>     headers:
>       Access-Control-Allow-Headers:
>       - origin, content-type, accept, Authorization, authorization
>       Access-Control-Allow-Methods:
>       - GET, OPTIONS
>       Access-Control-Allow-Origin:
>       - '*'
>       Access-Control-Max-Age:
>       - '3600'
>       Connection:
>       - close
>       Content-Length:
>       - '146'
>       Content-Type:
>       - application/json;charset=UTF-8
>       Date:
>       - Thu, 12 Aug 2021 19:14:33 GMT
>       Server:
>       - nginx/1.14.0 (Ubuntu)
>     status:
>       code: 200
>       message: ''
> version: 1
> ```
> 
> It looks like it didn't mock it properly?

Captura de ecrã 2021-08-19, às 19 14 56

it works locally but not here

**Comment by jmaslek** [comment]
> it works locally but not here

So I am lost then.  Not completely sure exactly how the VCR works, but it should not be dependent on the API or if there is a key.  It would make sense to assert that API Key Not found was in the response, but doesn't;t seem like it does that,

File: tests/cassettes/test_dd/test_financial_model/test_rating.yaml [hunk 1]
```diff
@@ -36,4 +36,82 @@ interactions:
...
+      - close
+      Host:
+      - financialmodelingprep.com
+      User-Agent:
+      - Python-urllib/3.8
+    method: GET
+    uri: https://financialmodelingprep.com/api/v3/historical-rating/PLTR?apikey=845b22316eeec4168d249976bba94330
+  response:
+    body:
+      string: "[ {\n  \"symbol\" : \"PLTR\",\n  \"date\" : \"2021-08-18\",\n  \"rating\"
```
Comment (DidierRLopes) [inline tests/cassettes/test_dd/test_financial_model/test_rating.yaml:52]: @jmaslek  if I used VCR correctly, which according to this it seems I did, it shouldn't fail.

Comment (jmaslek) [reply tests/cassettes/test_dd/test_financial_model/test_rating.yaml:52]: my bad.  Didn't see you committed the new yaml

Comment (DidierRLopes) [reply tests/cassettes/test_dd/test_financial_model/test_rating.yaml:52]: I didn't explain myself properly I think.

But is still failing not sure why

**Review by aia — APPROVED** [review]

File: gamestonk_terminal/stocks/due_diligence/business_insider_model.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+
+def get_estimates(ticker: str) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    """Get analysts' estimates for a given ticker. [Source: Business Insider]
+
+    Parameters
+    ----------
+    ticker : str
+        Ticker to get analysts' estimates
+    """
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/business_insider_model.py:69]: no returns doc

File: gamestonk_terminal/stocks/due_diligence/business_insider_view.py [hunk 3]
```diff
@@ -1,99 +1,66 @@
...
-        df_analyst_data = pd.DataFrame.from_dict(d_analyst_data["Markers"])  # type: ignore
-        df_analyst_data = df_analyst_data[
-            ["DateLabel", "Company", "InternalRating", "PriceTarget"]
-        ]
-        df_analyst_data.columns = ["Date", "Company", "Rating", "Price Target"]
-        # df_analyst_data
-        df_analyst_data["Rating"].replace(
-            {"gut": "BUY", "neutral": "HOLD", "schlecht": "SELL"}, inplace=True
+    if raw:
+        pd.set_option("display.max_colwidth", None)
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/business_insider_view.py:51]: why is this necessary with tabulate?

File: gamestonk_terminal/stocks/due_diligence/business_insider_view.py [hunk 4]
```diff
@@ -116,7 +83,7 @@ def price_target_from_analysts(
         plt.title(f"{ticker} (Time Series) and Price Target")
         plt.xlim(stock.index[0], stock.index[-1])
         plt.xlabel("Time")
-        plt.ylabel("Share Price ($)")
+        plt.ylabel("Share Price")
         plt.grid(b=True, which="major", color="#666666", linestyle="-")
         plt.minorticks_on()
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/business_insider_view.py:88]: can we lose the minorticks?  I have been dropping them as my personal preference is they make the plots less nice

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 5]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+        parser = argparse.ArgumentParser(
+            add_help=False,
+            prog="analyst",
+            description="""
+                Print analyst prices and ratings of the company. The following fields are expected:
+                date, analyst, category, price from, price to, and rating. [Source: Finviz]
+            """,
+        )
+        parser.add_argument(
+            "-n",
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:170]: thought -n was for number.

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 6]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+            "--no_color",
+            action="store_false",
+            dest="no_color",
+            default=True,
+            help="Remove coloring",
+        )
+        try:
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:177]: export

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 7]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+
+    def call_rot(self, other_args: List[str]):
+        """Process rot command"""
+        parser = argparse.ArgumentParser(
+            add_help=False,
+            prog="rot",
+            description="""
+                Rating over time. [Source: Finnhub]
+            """,
+        )
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:272]: export and (maybe) raw?

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 8]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+        parser.add_argument(
+            "-n",
+            "--num",
+            action="store",
+            dest="n_num",
+            type=check_positive,
+            default=10,
+            help="number of regions to plot that show highest interest.",
+        )
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:302]: export?

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 9]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+            if other_args:
+                if "-" not in other_args[0]:
+                    other_args.insert(0, "-n")
+
+            ns_parser = parse_known_args_and_warn(parser, other_args)
+            if not ns_parser:
+                return
+
+            fmp_view.rating(
+                ticker=self.ticker,
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:314]: Surprised no color in the table :P

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 10]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+        )
+        parser.add_argument(
+            "-n",
+            "--num",
+            action="store",
+            dest="n_num",
+            type=check_positive,
+            default=5,
+            help="number of latest SEC filings.",
+        )
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:339]: export?

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 11]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+                if "-" not in other_args[0]:
+                    other_args.insert(0, "-n")
+
+            ns_parser = parse_known_args_and_warn(parser, other_args)
+            if not ns_parser:
+                return
+
+            marketwatch_view.sec_filings(
+                ticker=self.ticker,
+                num=ns_parser.n_num,
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:352]: in the view - could we clip the link to just be the docid ?

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 12]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+            print(e, "\n")
 
     def call_supplier(self, other_args: List[str]):
         """Process supplier command"""
-        csimarket_view.suppliers(self.ticker, other_args)
+        parser = argparse.ArgumentParser(
+            prog="supplier",
+            add_help=False,
+            description="List of suppliers from ticker provided. [Source: CSIMarket]",
+        )
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:364]: export?  You could save the list as DataFrame(string.split('))

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py [hunk 13]
```diff
@@ -148,52 +158,251 @@ def call_quit(self, _):
...
+            print(e, "\n")
 
     def call_customer(self, other_args: List[str]):
         """Process customer command"""
-        csimarket_view.customers(self.ticker, other_args)
+        parser = argparse.ArgumentParser(
+            prog="customer",
+            add_help=False,
+            description="List of customers from ticker provided. [Source: CSIMarket]",
+        )
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/dd_controller.py:381]: same export as supplier

File: gamestonk_terminal/stocks/due_diligence/finnhub_model.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+    Returns
+    -------
+    pd.DataFrame
+        Get dataframe with ratings
+    """
+    response = requests.get(
+        f"https://finnhub.io/api/v1/stock/recommendation?symbol={ticker}&token={cfg.API_FINNHUB_KEY}"
+    )
+    if response.status_code == 200:
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/finnhub_model.py:25]: should we print out the status code if its not 200 (more of a general question for all commands everywhere)

File: gamestonk_terminal/stocks/due_diligence/finviz_view.py [hunk 15]
```diff
@@ -35,130 +30,53 @@ def category_color_red_green(val: str) -> str:
...
-        action="store",
-        dest="n_color",
-        type=int,
-        choices=[0, 1],
-        default=1,
-        help="Add / remove color",
+    df_fa = finviz_model.get_analyst_data(ticker)
+
+    if not no_color:
+        df_fa["category"] = df_fa["category"].apply(category_color_red_green)
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/finviz_view.py:69]: I don't get color with or without the flag

File: gamestonk_terminal/stocks/due_diligence/marketwatch_model.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    ticker : str
+        Stock ticker
+
+    Returns
+    -------
+    df_financials : pd.DataFrame
+        SEC filings data
+    """
+
+    pd.set_option("display.max_colwidth", None)
```
Comment (jmaslek) [inline gamestonk_terminal/stocks/due_diligence/marketwatch_model.py:26]: is this needed (especially in model)

File: gamestonk_terminal/stocks/due_diligence/business_insider_view.py (hunk 3 above)
Comment (DidierRLopes) [reply gamestonk_terminal/stocks/due_diligence/business_insider_view.py:51]: Shouldn't be! Going to fix it

File: gamestonk_terminal/stocks/due_diligence/business_insider_view.py (hunk 4 above)
Comment (DidierRLopes) [reply gamestonk_terminal/stocks/due_diligence/business_insider_view.py:88]: I thought it looked nice :( will get rid of it

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py (hunk 5 above)
Comment (DidierRLopes) [reply gamestonk_terminal/stocks/due_diligence/dd_controller.py:170]: Yup, good call. Didn't even notice

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py (hunk 7 above)
Comment (DidierRLopes) [reply gamestonk_terminal/stocks/due_diligence/dd_controller.py:272]: On it

File: gamestonk_terminal/stocks/due_diligence/dd_controller.py (hunk 11 above)
Comment (DidierRLopes) [reply gamestonk_terminal/stocks/due_diligence/dd_controller.py:352]: I prefer the whole link because it goes direct to the document
